services/repository/src/main/java/org/sagebionetworks/auth/filter/OAuthClientAuthFilter.java

```diff
@@ -1,91 +1,128 @@
+import org.sagebionetworks.StackConfiguration;
+import org.sagebionetworks.auth.HttpAuthUtil;
+import org.sagebionetworks.auth.UserNameAndPassword;
+import org.sagebionetworks.authutil.ModHttpServletRequest;
+import org.sagebionetworks.cloudwatch.Consumer;
+import org.sagebionetworks.repo.manager.oauth.OAuthClientManager;
+import org.sagebionetworks.repo.model.AuthorizationConstants;
+import org.sagebionetworks.repo.model.oauth.OAuthClientIdAndSecret;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * Implements the client_secret_basic and client_secret_post protocols for authenticating an OAuth client
+ * See: https://tools.ietf.org/html/rfc6749#section-2.3.1
+ */
+@Component("oauthClientAuthFilter")
+public class OAuthClientAuthFilter implements Filter {
+	private static final String INVALID_CREDENTIALS_MSG = "OAuth Client ID and secret must be passed via Basic Authentication or as request parameters. Credentials are missing or invalid.";
+	private static final String CLIENT_ID_PARAM_NAME = "client_id";
+	private static final String CLIENT_SECRET_PARAM_NAME = "client_secret";
+
+	private StackConfiguration config;
+
+	private Consumer consumer;
+
+	private OAuthClientManager oauthClientManager;
+
+	private static final boolean REPORT_BAD_CREDENTIALS_METRIC = true;
+	
+	@Autowired
+	public OAuthClientAuthFilter(StackConfiguration config, Consumer consumer, OAuthClientManager oauthClientManager) {
+		this.config=config;
+		this.consumer=consumer;
+		this.oauthClientManager=oauthClientManager;
+	}
+
+	private boolean validCredentials(UserNameAndPassword credentials) {
+		OAuthClientIdAndSecret clientCreds = new OAuthClientIdAndSecret();
+
+		clientCreds.setClient_id(credentials.getUserName());
+		clientCreds.setClient_secret(credentials.getPassword());
+
+		return this.oauthClientManager.validateClientCredentials(clientCreds);
+	}
+	
+	@Override
+	public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
+			throws IOException, ServletException {
+
+		if (!(request instanceof HttpServletRequest) || !(response instanceof HttpServletResponse)) {
+			throw new ServletException("Only HTTP requests are supported");
+		}
+
+		HttpServletRequest httpRequest = (HttpServletRequest) request;
+		HttpServletResponse httpResponse = (HttpServletResponse) response;
+
+		Optional<UserNameAndPassword> credentials;
+		
+		FilterHelper filterHelper = new FilterHelper(config, consumer);
+
+		try {
+			credentials = HttpAuthUtil.getBasicAuthenticationCredentials(httpRequest);
+		} catch (IllegalArgumentException e) {
+			filterHelper.rejectRequest(REPORT_BAD_CREDENTIALS_METRIC, httpResponse, e);
+			return;
+		}
+		
+		// if credentials are not in the authorization header (i.e. passed by client_secret_basic),
+		// the might be passed as request parameters (i.e. passed by client_secret_post)
+		if (!credentials.isPresent()) {

```

Comment:
Might be worth checking that it is an actual POST or that the params are not URL params